content/vendor/guides/persistent-datastores.md

```diff
@@ -0,0 +1,940 @@
+```
+
+
+after deploying this, you should see all fields in the secret:
+
+```text
+$ kubectl get secret postgres -o yaml
+apiVersion: v1
+data:
+  DB_HOST: ZmFrZQ==
+  DB_NAME: ZmFrZQ==
+  DB_PASSWORD: ZXh0cmEgZmFrZQ==
+  DB_PORT: NTQzMjE=
+  DB_USER: ZmFrZQ==
+kind: Secret
+# ...snip...
+```
+
+We can also print the environment in our sample app to verify all the values are piped properly:
+
+```text
+$ kubectl exec $(kubectl get pod -lapp=pg-consumer -o jsonpath='{.items[0].metadata.name}' ) -- /bin/sh -c 'printenv | grep DB_'
+DB_PORT=54321
+DB_NAME=fake
+DB_PASSWORD=extra fake
+DB_HOST=fake
+DB_USER=fake
+```
+
+### Testing Config Changes
+
+Now let's make some changes to the database credentials. In this case, we'll use a postgres database provisioned in Amazon RDS, but you can use any external database. To Start, head to the config screen and input your values:
+
+![Real Postgres Values](/images/guides/kots/real-postgres-values.png)
+
+Let's save and apply this config and check in our pod again:
+
+```text
+$ kubectl exec $(kubectl get pod -lapp=pg-consumer -o jsonpath='{.items[0].metadata.name}' ) -- /bin/sh -c 'printenv | grep DB_'
+DB_PORT=54321
+DB_NAME=fake
+DB_PASSWORD=extra fake
+DB_HOST=fake
+DB_USER=fake
+```
+
+Uh oh, It appears that our values did not get updated! If you've worked with Secrets before, you may know that there's a [long-standing issue in Kubernetes](https://github.com/kubernetes/kubernetes/issues/22368) where pods that load config from Secrets or ConfigMaps won't automatically restart when underlying config is changed. There are some tricks to make this works, and in the next step we'll implement one of them, but for now we can delete the pod to verify that the configuration is being piped through to our sample application:
+
+```text
+$ kubectl delete pod -lapp=pg-consumer
+pod "pg-consumer-6df9d5d7fd-bd5z6"" deleted
+```
+
+If the pod is crashlooping, you might need to add `--force --grace-period 0` to force delete it. In either case, once a new pod starts, we should now see it loading the correct config:
+
+```text
+$ kubectl exec $(kubectl get pod -lapp=pg-consumer -o jsonpath='{.items[0].metadata.name}' ) -- /bin/sh -c 'printenv | grep DB_'
+DB_PORT=5432
+DB_NAME=postgres
+DB_PASSWORD=<redacted>
+DB_HOST=10.128.0.12
+DB_USER=postgres
+```
+
+### Trigerring restarts on changes
+
+In order to automate this restart on changes, we're going to use a hash of all database parameters to trigger a rolling update whenever database parameters are changed. We'll use a `hidden`, `readonly` field to store this in our config screen:
+
+```yaml
+        - name: external_postgres_confighash
+          hidden: true
+          readonly: true
+          type: text
+          value: '{{repl (sha256 (print (ConfigOption "external_postgres_host") (ConfigOption "external_postgres_port") (ConfigOption "external_postgres_user") (ConfigOption "external_postgres_password") (ConfigOption "external_postgres_db") ))}}'
```

Comment:
This will fail to deploy due to missing function. Change to sha256sum and you're good. 

![image](https://user-images.githubusercontent.com/8466736/79285617-f0d3b800-7e72-11ea-9724-ecf850cd0708.png)

Reply:
thanks, had it fixed in kotsapp but forgot to port it back. 👍